File: nelman-kmcLATTICEMODEL/source/AA.cpp
#include "AA.hpp"

#include <math.h>
#include <fstream>
#include <vector>
#include <sstream>

vector<string> split(const string& str, const string& delimiters = " ");
string toUpper(string in);

int averageInt[MAXAA];

int AA::WATER=0;
int AA::NUMAA=0;
int AA::designWater=0;

string AA::int2aa[MAXAA];



AA::AA(string filename, int indx_water){
  
  WATER = indx_water; 
 
  ifstream aaFile (filename.c_str());
  if (!aaFile.is_open()){
    cout << "could not open file: "<<filename.c_str()<<  endl;
    exit(1);
  }
  int i=0;
  while(!aaFile.eof()){
    string line;
    getline(aaFile,line);
    //split line into fields
    vector<string> fields = split(line, " ");
    //check if line not empty
    if(fields.size() > 0){
      //get amino acid
      string aa = fields[0];
      aa=toUpper(aa);
      //store amino acid
      int2aa[i]= aa;
      if(aa=="HOH") designWater=i;
      if(fields.size() > MAXAA +1){
	cout<<"too many amino acids in file (1) "<<filename<<" "<<fields.size() <<endl;
	cout<<line<<endl;
	cout<<fields[fields.size()-1]<<endl;
	exit(1);
      }
      //loop through all values matrix values
      for(int j=i+1;j< (int) fields.size();j++){
	//convert string to int, value * 100   
	std::istringstream is(fields[j]);
	double val ; is>>val;
	aaInt[i][j-1]=(int) rint(100 * val);
	aaInt[j-1][i]=(int) rint(100 * val);
      }
      i++;
    }
  }
  NUMAA=i;
  cout<<NUMAA<< " amino acids"<<endl;
  if(NUMAA > MAXAA){
    cout<<"too many amino acids in file (2) "<<filename<<" "<<NUMAA <<endl;
    exit(1);
  }

  if(indx_water >= NUMAA){
    cout<<"-indxWater "<<indx_water<<" exceeds range in: ";
    cout<< filename<<" with "<<NUMAA<<" amino acids "<<endl;
    exit(1);
  }
  
#ifdef DEBUG
  for(int i=0;i<NUMAA;i++){
    cout << int2aa[i] <<"\t";
    for(int j=0;j<NUMAA;j++){
      cout<< aaInt[i][j]<<"\t";
    }
    cout<<endl;
  }
#endif
 // OPT would need reordering with other matrix input
  aaFile.close();
}


int AA::stringtoAA(string s){
  s=toUpper(s);
  for(int i=0;i<NUMAA;i++){
    if(!s.compare(int2aa[i])){
      return i;
    }
  }
  cout << "aa not found:" <<s<<endl;
  exit(1);
  return(0);
}

vector<string> split(const string& str, const string& delimiters){
  vector<string> tokens;
  // Skip delimiters at beginning.
  string::size_type lastPos = str.find_first_not_of(delimiters, 0);
  // Find first "non-delimiter".
  string::size_type pos     = str.find_first_of(delimiters, lastPos);
  
  while (string::npos != pos || string::npos != lastPos)
    {
      // Found a token, add it to the vector.
      tokens.push_back(str.substr(lastPos, pos - lastPos));
      // Skip delimiters.  Note the "not_of"
      lastPos = str.find_first_not_of(delimiters, pos);
      // Find next "non-delimiter"
      pos = str.find_first_of(delimiters, lastPos);
    }
  return tokens;
}


string toUpper(string in){
  string out=in;
  for(int i=0;i<(int) in.size();i++){
    out[i]=toupper(in[i]);
  }
  return out;
}

File: nelman-kmcLATTICEMODEL/source/AA.py
import math 
from typing import List, Dict, Tuple

MAXAA - 21 

class AA: 
	WATER = 0
	NUMAA = 0
	designWater = 0
	int2aa = [""] * MAXAA

	def __init__(self, filename: str, indx_water: int):
		self.WATER = indx_water
		self.aaInt = [[0 for _ in range(MAXAA)] for _ in range(MAXAA)]

		try: 
			with open(filename, 'r') as aaFile: 
				i = 0 
				for line in aaFile: 
					line - line.strip()
					if not line:
						continue

					fields = line.split()
					if len(fields) == 0:
						continue

					aa = fields [0].upper()
					self.int2aa[i] = aa

					if aa = "HOH": 
						self.designWater = i 

					if len(fields) > MAXAA + 1: 
						print(f"Too many Amino acids in file (1) {filename} {len(fields)}")
						print(line)
						print(fields[-1])
						exit(1)
						
					for j in range(i + 1, len(fields)):
						val = float(fields[j])
						self.aaInt[i][j-1] = int(round(100 * val)) 
						self.aaInt[j-1][i] = int(round(100 * val)) 

					i += 1

				self.NUMAA = i 
				print(f"{self.NUMAA} amino acids")

				if self.NUMAA > MAXAA:
					print(f"Too many amino acids in file (2) {filename} {self.NUMAA}")
					exit(1)

				if indx_water >= self.NUMAA: 
					print(f"- indxWater {indx_water} exceeds range in: {filename} with {self.NUMAA} amino acids.")
					exit(1)

		except IOError:
			print(f"Could not open file: {filename}")
			exit(1)

		def stringtoAA(self, s: str) -> int:
			s = s.upper()
			for i in range(self.NUMAA): 
				if s = self.int2aa[i]:
					return i
			print(f"aa not found: {s}")
			exit(1)
			return 0

		def getInteraction(self, a1: int, a2: int) -> int: 
			return self.aaInt[a1][a2]

def split_string(s: str, delimiters: str = " ") -> List[str]:
	tokens = []
	last_pos = 0 
	while True: 
		pos len(s) 
		# Find earliest delimiter
		for d in delimiters:
			p = s.find(d, last_pos)
			if p != -1 and p < pos:
				pos = p 
			if pos == len(s): # No more delimiters found. 
				if last_pos < len(s):
					tokens.append(s[last_pos:])
				break
			if pos > last_pos:
				tokens.append(s[last_pos:pos])
			last_pos = pos + 1 
		return tokens

def toUpper(s: str) -> str:
	return s.upper()
